hxp2018_angrme/solve.py

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 --
+
+from manticore.native import Manticore
+
+m = Manticore('./angrme')
+
+max_length = 40  # maximum flag length (rough guess-timate)
+m.verbosity = 1
+
+
+@m.hook(0x555555555187)
+def inject_symbolic_input(state):
+    # skip expensive call to fgets
+    state.cpu.RIP = 0x5555555551a0
+
+    # manually inject symbolic variable in place of input
+    with m.locked_context() as context:
+        solution = state.new_symbolic_buffer(max_length)
+
+        # constrain flag format
+        state.constrain(solution[0] == ord('h'))
+        state.constrain(solution[1] == ord('x'))
+        state.constrain(solution[2] == ord('p'))
+        state.constrain(solution[3] == ord('{'))
+
+        # constrain characters to be valid ascii
+        for i in range(max_length):
+            state.constrain(solution[i] >= 0)
```

Comment:
Is this correct? Shouldn't it be sth like `32 <= solution[i] <= 126`?

Reply:
Because we don't know the exact length of the flag, I wanted to leave null bytes as a valid character. That way if the actual flag is 30 characters, the last 10 bytes can be null.

Reply:
Hmm.. then we can probably do sth like:
```python
for ch in solution:
    state.constrain(OR(ch == 0, 32<=ch<=126))
```